core/src/hash/hasher.rs

```diff
@@ -0,0 +1,360 @@
+pub trait FinishInto<Output> {
+    fn finish_into(&self, output: &mut Output);
+}
+
+/// Hash functions that can return their hash results as an output buffer
+/// described by the `Output` type.
+///
+/// # Note
+///
+/// - Do not manually implement this trait since it is automatically implemented
+///   as `Finish<O>` for all `T: FinishInto<O>`.
+/// - This is an internal trait used to deduplicate some implementations.
+pub trait Finish<Output> {
+    fn finish(&self) -> Output;
+}
+
+impl<T, A, O> Finish<O> for CryptoHasher<T, A>
+where
+    Self: FinishInto<O>,
+    T: markers::Sealed,
+    A: Accumulator,
+    O: AsMut<[u8]> + Default,
+{
+    fn finish(&self) -> O {
+        let mut output = <O as Default>::default();
+        <Self as FinishInto<O>>::finish_into(self, &mut output);
+        output
+    }
+}
+
+impl<A> FinishInto<[u8; 32]> for Sha2x256Hasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 32]) {
+        super::sha2_256_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 32]> for Keccakx256Hasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 32]) {
+        super::keccak_256_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 32]> for Blake2x256Hasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 32]) {
+        super::blake2_256_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 16]> for Blake2x128Hasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 16]) {
+        super::blake2_128_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 32]> for TwoxHasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 32]) {
+        super::twox_256_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 16]> for TwoxHasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 16]) {
+        super::twox_128_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 8]> for TwoxHasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 8]) {
+        super::twox_64_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+/// Crypto hash functions that allow to return their hash as `u64`.
+///
+/// # Note
+///
+/// - This is an internal trait used to deduplicate some implementations.
+/// - This is a compatibility function for direct usage of the supported
+///   crypto hash functions via [`Hash`](`core::hash::Hash`) trait.
+pub trait FinishU64 {
+    /// Returns the `u64` result of the accumulated hash.
+    fn finish(&self) -> u64;
+}
+
+impl<A> FinishU64 for Sha2x256Hasher<A>
+where
+    A: Accumulator,
+{
+    fn finish(&self) -> u64 {
+        let [h0, h1, h2, h3, h4, h5, h6, h7, ..] =
```

Comment:
this looks like it needs a helper function

with hashes, it is better to never to shuffle bytes individually, since errors are sometimes hard to catch - hash bytes are uniform (look random)

Reply:
How to better extract the first 8 bytes out of a byte slice?
Note that
```
let first_8_bytes: [u8; 8] = <Self as Finish<[u8; 32]>>::finish(self)[0..8];
```
Doesn't work since Rust cannot infer `[u8; 8]` type here, just a `[u8]` so you'd need to unsafely cast the type or use some crate that does the job for you but then again you don't want unnecessary dependencies.

Reply:
Yeah, it's a known Rust problem. However, you can just add this:

u64_from_slice_first_8_bytes(bytes: &[u8]) -> u64 {
   let mut first8bytes = [0u8; 8];
   firstBytes[..].copy_from_slice(&bytes[..8];
   u64::from_le_bytes(firstBytes)
}

and use it 4 times here 🤷‍♀️

Reply:
Ah that's what you meant. xD Yeah of course.

Reply:
done